### PR TITLE
dev: munge PATH to remove ccache entries when calling into Bazel

### PR DESCRIPTION
**Pull request by rail**
Previously, if `ccache` is setup on a developer computer and the
`ccache` directory was at the head of `PATH`, it lead to compile errors,
because `ccache` tried to write to a directory which is outside of
bazel's sandbox.

This patch detects if a compiler is symlinked to `ccache` and removes
the detected directory from `PATH`.

Release note: None

### CODE REVIEW

**Comment by rail** [comment]
Fixes https://github.com/cockroachdb/dev/issues/11

File: pkg/cmd/dev/util.go [hunk 1]
```diff
@@ -105,11 +107,74 @@ func addCommonTestFlags(cmd *cobra.Command) {
 	cmd.Flags().Duration(timeoutFlag, 0*time.Minute, "timeout for test")
 }
 
-func ensureBinaryInPath(bin string) error {
+func ensureBinaryInPath(dev *dev, bin string) error {
```
Comment (rickystewart) [inline pkg/cmd/dev/util.go:110]: Nit: Consider making this a method on `*dev` instead of passing it as a parameter

File: pkg/cmd/dev/util.go [hunk 2]
```diff
@@ -105,11 +107,74 @@ func addCommonTestFlags(cmd *cobra.Command) {
...
+		err = setupPathReal(dev)
+	})
+	return err
+}
+
+// setupPathReal uses a list of known compiler names to check if they are
+// symlinks to `ccache`.
+func setupPathReal(dev *dev) error {
+	knownCompilers := []string{"cc", "gcc", "c++", "g++", "clang", "clang++"}
+	// datadriven uses Fprintln() to construct commands, adding '\n' to the end.
```
Comment (rickystewart) [inline pkg/cmd/dev/util.go:135]: Pretty unfortunate that this is necessary. Thanks for adding the comment to clarify though.

File: pkg/cmd/dev/util.go [hunk 3]
```diff
@@ -105,11 +107,74 @@ func addCommonTestFlags(cmd *cobra.Command) {
...
+		for _, dir := range pathEntries {
+			if compilerDir != dir {
+				newPathEntries = append(newPathEntries, dir)
+			}
+		}
+		newPath := strings.Join(newPathEntries, string(os.PathListSeparator))
+		if origPath == newPath {
+			log.Printf("WARNING: PATH did not change: %s", origPath)
+		}
+		log.Printf("INFO: removed ccache directory from PATH. New PATH: %s", newPath)
```
Comment (rickystewart) [inline pkg/cmd/dev/util.go:172]: Nit: This should be silent under most circumstances. Printing this out every time someone invokes `dev` is not helpful. If you still want the message you can do `if debugVar { log.Printf(...) }`

**Review by rickystewart — APPROVED** [review]
LGTM pending comments

File: pkg/cmd/dev/util.go (hunk 1 above)
Comment (rail) [reply pkg/cmd/dev/util.go]: Done

File: pkg/cmd/dev/util.go (hunk 3 above)
Comment (rail) [reply pkg/cmd/dev/util.go]: Removed.

**Comment by rail** [comment]
bors r=rickystewart

**Pull request merged**